SPEC.md

```diff
@@ -0,0 +1,334 @@
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
+schema:
+
+```yaml
+type: object
+properties:
+  message:
+    type: string
+  details:
+    type: object
+    properties:
+      metadata:
+        type: object
+        additionalProperties:
+          type: string
+        description: |
+          String to string mapping, may contain information for decoding the data field.
+      data:
+        type: string
+        description: |
+          Arbitrary string-encoded data; use encodings like base64 for binary data transmission.
+```
+
+### OperationStarted
+
+An `OperationStarted` object indicates the commencement of an asynchronous operation. This object MUST follow the given
+JSON schema:
+
+```yaml
+type: object
+properties:
+  operationId:
+    type: string
+    description: |
+      An identifier for referencing this operation. If provided in the start request, it is echoed back.
+  callbackUrlSupported:
+    type: boolean
+    description: |
+      If the request specified a callback_url and the handler supports callbacks for this operation, this flag will be
+      set.
+      It is up to the caller to decided how to get the outcome of this call in case the handler does not support
+      callbacks.
+```
+
+### OperationInfo
+
+The `OperationInfo` object, retrieved from the get operation info API, MUST adhere to the given schema:
+
+```yaml
+type: object
+properties:
+  state:
+    enum:
+      - succeeded
+      - failed
+      - canceled
+      - running
+    description: |
+      Describes the current state of an operation.
+```
+
+## Endpoint Descriptions
+
+### Start Operation
+
+Start an arbirary length operation.
+The response of the operation may be delivered synchronously (inline), or asynchronously, via a provided callback or the
+[Get Operation Result](#get-operation-result) endpoint.
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}`
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
```

Comment:
Is the operation_id always known to the client when calling Start? I saw your comment about handler generated ids, this is probably affected...

Reply:
This is in the process of changing. There are internal disagreements here, but suffice to say we're probably removing the ability for a client to identify an operation at first. I disagree with it (as I do for anything that makes things harder for callers such as having to store an identifier on a result to reference an operation), but oh well.